http-gateway/web/src/containers/things/utils.js

```diff
@@ -104,6 +115,70 @@ export const updateThingsDataStatus = (data, { deviceId, status }) => {
   })
 }
 
+/** Tree Structure utilities **/
+// Shout out to @oskarbauer for creating this script :)
+
+// A recursive function which "densify" the subRows
+const deDensisfy = objectToDeDensify => {
+  const { href, ...rest } = objectToDeDensify
+
+  const keys = Object.keys(rest)
+  return keys.map(thisKey => {
+    const value = objectToDeDensify[thisKey]
+    if (value.subRows) {
+      value.subRows = deDensisfy(value.subRows)
+    }
+    return value
+  })
+}
+
+// A recursive function for creating a tree structure from the href attribute
+const addItem = (objToAddTo, item, position) => {
+  const { href, ...rest } = item
+  const parts = href.split('/')
+  const isLast = position === parts.length - 1
+  position = position + 1
+  const key = `/${parts.slice(1, position).join('/')}/`
+
+  if (isLast) {
+    objToAddTo[key] = { ...objToAddTo[key], ...rest, href: key }
+  } else {
+    objToAddTo[key] = {
+      ...objToAddTo[key],
+      ...(key === href ? rest : {}),
+      href: key,
+      subRows: { ...(objToAddTo[key]?.subRows || {}) }, // subRows is the next level in the tree structure
+    }
+    // Go deeper with recursion
+    addItem(objToAddTo[key].subRows, item, position)
+  }
+}
+
+export const createNestedResourceData = data => {
+  // Always construct the objects from scratch
+  let firstSwipe = {}
+  if (data) {
+    data.forEach(item => {
+      addItem(firstSwipe, item, 1)
+    })
+  }
+  const output = deDensisfy(firstSwipe)

```

Comment:
//  Then take the object structure and output the tree scructure

Reply:
Thanks.